[SPARK-21798]: No config to replace deprecated SPARK_CLASSPATH config for launching daemons like History Server

pgandhi999 opened the pull request:
History Server Launch uses SparkClassCommandBuilder for launching the server. It is observed that SPARK_CLASSPATH has been removed and deprecated. For spark-submit this takes a different route and spark.driver.extraClasspath takes care of specifying additional jars in the classpath that were previously specified in the SPARK_CLASSPATH. Right now the only way specify the additional jars for launching daemons such as history server is using SPARK_DIST_CLASSPATH (https://spark.apache.org/docs/latest/hadoop-provided.html) but this I presume is a distribution classpath. It would be nice to have a similar config like spark.driver.extraClasspath for launching daemons similar to history server.

Added new environment variable SPARK_DAEMON_CLASSPATH to set classpath for launching daemons. Tested and verified for History Server and Standalone Mode.

## How was this patch tested?
Initially, history server start script would fail for the reason being that it could not find the required jars for launching the server in the java classpath. Same was true for running Master and Worker in standalone mode. By adding the environment variable SPARK_DAEMON_CLASSPATH to the java classpath, both the daemons(History Server, Standalone daemons) are starting up and running.

Commit:
[SPARK-21503]: Fixed the issue

Added the case ExecutorLostFailure which was previously not there, thus, the default case would be executed in which case, task would be marked as completed.

Commit:
[SPARK-21503][UI]: Adding changes as per comments

pgandhi999 committed:
Merge pull request #1 from apache/master

Apache Spark Pull Request - July 26, 2017

Commit:
[SPARK-21541]: Spark Logs show incorrect job status for a job that does not create SparkContext

Added a flag to check whether user has initialized Spark Context. If it is true, then we let Application Master unregister with Resource Manager else we do not.

Commit:
Revert "[SPARK-21541]: Spark Logs show incorrect job status for a job that does not create SparkContext"

This reverts commit f454c8933e07967548095e068063bd313ae4845c.

"Merged another issue to this one by mistake"

pgandhi999 committed:
Merge pull request #2 from apache/master

Spark - July 28, 2017

Commit:
[SPARK-21503]- Making Changes as per comments

[SPARK-21503]- Making Changes as per comments: Removed match case statement and replaced it with an if clause.

pgandhi999 committed:
Merge pull request #4 from apache/master

Spark - August 1, 2017

Commit:
[SPARK-21503]: Reverting Unit Test Code

[SPARK-21503]: Reverting Unit Test Code - Not needed.

pgandhi999 committed:
Merge pull request #5 from apache/master

SPARK - August 24, 2017

Commit:
[SPARK-21798]: No config to replace deprecated SPARK_CLASSPATH config for launching daemons like History Server

Adding new env variable SPARK_DAEMON_CLASSPATH to set classpath for launching daemons. Tested and verified for History Server and Standalone Mode.

AmplabJenkins commented:
Can one of the admins verify this patch?

jerryshao commented on launcher/src/main/java/org/apache/spark/launcher/AbstractCommandBuilder.java:
```diff
@@ -136,7 +136,8 @@ void addOptionString(List<String> cmd, String options) {
     Set<String> cp = new LinkedHashSet<>();
     addToClassPath(cp, appClassPath);
-
+    addToClassPath(cp, getenv("SPARK_DAEMON_CLASSPATH"));
```
I think it would be better to only support this env variable for daemon process like HistoryServer, ExternalShuffleService and others, like what we did for `SPARK_DAEMON_JAVA_OPTS`. Currently with your fix normal Spark application will also honor this env variable.

pgandhi999 replied:
I agree with you on this. I have made the necessary changes.

Commit:
[SPARK-21798]: No config to replace deprecated SPARK_CLASSPATH config for launching daemons like History Server

Reverted the previous code change and added the environment variable SPARK_DAEMON_CLASSPATH only for launching daemon processes.

tgravescs commented:
ok to test

SparkQA commented:
**[Test build #81133 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81133/testReport)** for PR 19047 at commit [`e421a03`](https://github.com/apache/spark/commit/e421a03acbd410a835cf3117fe6592523dc649b5).

SparkQA commented:
**[Test build #81133 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81133/testReport)** for PR 19047 at commit [`e421a03`](https://github.com/apache/spark/commit/e421a03acbd410a835cf3117fe6592523dc649b5).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

tgravescs commented:
test this please

SparkQA commented:
**[Test build #81141 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81141/testReport)** for PR 19047 at commit [`e421a03`](https://github.com/apache/spark/commit/e421a03acbd410a835cf3117fe6592523dc649b5).

vanzin commented:
LGTM. I'd like to see these daemons start using normal Spark configs like the applications do, but that's a separate, larger change...

SparkQA commented:
**[Test build #81141 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81141/testReport)** for PR 19047 at commit [`e421a03`](https://github.com/apache/spark/commit/e421a03acbd410a835cf3117fe6592523dc649b5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81141/
Test PASSed.

tgravescs commented:
+1
